### ISSUE TITLE
Wrong type emitted when using type instead of interface

### ISSUE THREAD

**Issue by rubenlg** [angular/tsickle#451]
I found this while using firebase. Here is how they declare firebase.auth.UserCredential in their d.ts:

```typescript
  type UserCredential = {
    credential: firebase.auth.AuthCredential | null,
    user: firebase.User | null
  };
```

If I use that type like this:
```typescript
  /** @export */
  signInComplete(userCredential: firebase.auth.UserCredential) {
    if (!userCredential.user) {
      console.log(userCredential);
      throw Error('Credential with no user');
    }
 [... more code ...]
}
```

The emitted code is:

```javascript
    /**
     * @export
     * @param {{credential: (null|!firebase.auth.AuthCredential), user: (null|!firebase.User)}} userCredential
     * @return {void}
     */
    signInComplete(userCredential) {
        if (!userCredential.user) {
            console.log(userCredential);
            throw Error('Credential with no user');
        }
    [... more code...]
  }
```

The type is inlined, which lets the closure compiler rename `.user` freely, causing the exception to be thrown. Interestingly, the symbol is correctly emitted in the externs file:

```javascript
/** @typedef {{credential: (null|!AuthCredential), user: (null|!User)}} */
firebase.auth.UserCredential;
```

Tsickle should have been used the exported symbol instead, so that closure preserves the names of the properties.

**Comment by mprobst** [comment]
Using the exported symbol will not prevent renaming. You'd reasonably think that that's how Closure Compiler works, but it isn't ;-) AFAIK Closure will just avoid renaming properties that could match externs defined properties, regardless of whether you reference a type from externs (unless you're using advanced type based optimizations, and there are a bunch of additional flags that influence this...).

I think the problem why you see renaming happening is that Closure has a long standing bug where it renames properties from externs that only appear in typedefs. The fix is to use an `@record` interface (which is roughly the same on the TS side anyway).

Maybe tsickle should simply translate all typedefs to `@record` interfaces for externs to work around the issue?

**Comment by rubenlg** [comment]
Closure seems to be working fine with @typedef for me. I used it before moving to typescript and prevented renaming in -O ADVANCED.

I have another theory on what the problem might be here. Looks like tsickle is getting confused by namespaces when generating the externs file. For example, for firebase.d.ts, it produces the following externs.js:

```javascript
/**
 * @param {!auth.AuthCredential} credential
 * @return {!Promise<?>}
 */
firebase.User.prototype.link = function(credential) {};
[...]
/** @record @struct */
firebase.auth.AuthCredential = function() {};
[...]
/** @typedef {{credential: (null|!AuthCredential), user: (null|!User)}} */
firebase.auth.UserCredential;
```

On the first case, `auth.AuthCredential` is missing the first namespace token `firebase`. On the latter, which is the one from my previous comment, it misses both `firebase` and `auth`. That seems like a more likely reson for the bug. `AuthCredential` without namespace is never defined on that externs file, so closure might not know what that is?

To get the file `firebase.d.ts` you can just do `npm install firebase`, but in short, all usages of `AuthCredential` on that file are fully qualified, and the declaration is inside a `declare namespace firebase.auth` block.

I tried that with a smaller example, here is a toy `d.ts`:

```typescript
declare namespace root {
  interface Service {
    onEvent(callback: (ev: root.events.MyEvent) => void): void;
  }
  const service: Service;
}

declare namespace root.events {
  type MyEvent = {
    name: string,
    height: root.events.Height
  };

  interface Height {
    value: number;
    units: string;
  }
}
```

Here is the produced `externs.js` file:

```javascript
// externs from externs.d.ts:
/**
 * @externs
 * @suppress {duplicate}
 */
// NOTE: generated by tsickle, do not edit.
/** @const */
var root = {};

/** @record @struct */
root.Service = function() {};

/**
 * @param {function({name: string, height: !events.Height}): void} callback
 * @return {void}
 */
root.Service.prototype.onEvent = function(callback) {};
 /** @type {!Service} */
root.service;
/** @const */
root.events = {};

/** @typedef {{name: string, height: !Height}} */
root.events.MyEvent;

/** @record @struct */
root.events.Height = function() {};
 /** @type {number} */
root.events.Height.prototype.value;
 /** @type {string} */
root.events.Height.prototype.units;
```

Here is the compiler output when compiling with those externs:

```sh
tsickle --externs=externs.js
closure-compiler -O ADVANCED --warning_level VERBOSE --language_in ECMASCRIPT6 --language_out ECMASCRIPT5 test.js --externs=externs.js

externs.js:14: WARNING - Bad type annotation. Unknown type events.Height
 * @param {function({name: string, height: !events.Height}): void} callback
                                            ^

externs.js:18: WARNING - Bad type annotation. Unknown type Service
 /** @type {!Service} */
             ^

externs.js:23: WARNING - Bad type annotation. Unknown type Height
/** @typedef {{name: string, height: !Height}} */
```

**Comment by evmar** [comment]
Possibly related: https://github.com/angular/tsickle/issues/369

**Comment by rubenlg** [comment]
Seems indeed like the same issue. I'll close this bug. For others landing on this page while looking for answers, here is a workaround that did the trick for me. I just declared an interface and used that instead of the type on the .d.ts:

```typescript
declare interface FbUserCredential {
  credential: firebase.auth.AuthCredential | null,
  user: firebase.User | null
};
```

Now tsickle adds the right thing to the externs file, and closure doesn't rename the properties anymore.

Looking forward to the fix for #369 so I can drop the hack :)
